Monika-After-Story/MonikaModDev
# Added splitting hair layers to sprite chart

This makes it possible to have hair versions that are independent of clothes by allowing for hair objects that can be split between back and front layers. All existing hair styles can still be used by setting the `MASHair` property `split` to False for those hair objects. 

Also added additional accessory layer spots, and a little bit of optimization when removing accessories.

The new render order:
1. pre-acs - all accessories that should be rendered before everything else.
2. back-hair - the back portion of a hair object (only if split is True)
3. pre-body-acs - accessories between the back hair and before the body (only if split is True)
4. body - the body (not including arms if not leaning) (only if split is True)
5. pre-front-hair-acs - accessories between the body and the front portion of a hair object (only if split is True)
6. front-hair - the front portion of a hair object (only if split is True)
7. arms - the arms (only if not leaning) (only if split is True) 
8. mid-acs - accessories that should be rendered between the body and face
9. face - face expressions
10. pst-acs - accessories that should be rendered after everything else.

Also did another small optimization with loc_strings, so we dont need to pass position strings to all sub composites.


# new file requirements:

## hair
All generic hair sprites (ones that can be used for any clothing), should be placed in `mod_assets/monika/h/`. The naming scheme is:
```python
hair-{hair name}-{front/back}{-n}.png
```
Use `front` or `back` to denote a front layer or a back layer.

## body/clothes
Body types (or clothes) that are no longer attached to hair should follow this naming scheme:
These should still be placed in the `c/<clothes>` folder.
```python
# non-leaning
body-def{-n}.png
# leaning
body-leaning-{lean type}{-n}.png
```

Use the original naming scheme (with `torso`) if the body sprites have hairstyles baked into them.

# custom hair
I have added a new hair object called `mas_hair_custom`. this is a special hair object that simply has the `split` property set to True. This allows for costumes with baked hairstyles (like marisa and rin) to work even after all existing hairstyles get split. Clothes that have baked hairstyles shoudl set their hair maps to use `custom` for all hair styles.

# Testing
## regression
1. verify that sprites look still look okay.
2. open exp previewer and verify that the poses all look okay.
3. verify that the o31 costumes still look okay.
4. verify zooming still looks okay (even with accessories)

## hair layer test
Setup:
* For this, I recommend grabbing the ponytail layer that @kaido1224 posted on discord.
1. Make a folder in `mod_assets/monika` called `h`.
2. Copy over `torso-bun` from `mod_assets/monika/c/def/` into `mod_assets/monika/h`.
3. Copy the ponytail layer into `mod_assets/monika/h` as well.
4. In the `h` folder, do the following:
    1. Rename `torso-bun` (and its night version) into `hair-bun-front`
    2. Rename `ponytail` into `hair-bun-back`
5. In `mod_assets/monika/c/def/`, do the following:
    1. Make a copy of `torso-bun` and call it `body-def`. (also make a night version)
6. open `sprite-chart` and set the `split` property of `mas_hair_bun` to True.

Test:
1. launch game.
2. Open console and do `monika_chr.change_hair(mas_hair_bun)`. 
3. verify no crash.
4. Monika should have the ponytail, but also the hair coming down the right side of her face, which is part of the bun hairstyle.
5. Also verify that the arms are rendered over the front hair (aka you can see them move and shit).
6. Exp previewer should probably still work maybe.
